crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,327 @@
+        let nix: NixCommandLine = self.flox.nix(Default::default());
+
+        let locked_ref_command = runix::command::FlakeMetadata {
+            flake_ref: Some(self.publish_ref.clone().into_inner().into()),
+            ..Default::default()
+        };
+
+        locked_ref_command
+            .run_typed(&nix, &Default::default())
+            .map_err(|nix_err| PublishError::FlakeMetadata(self.publish_ref.clone(), nix_err))
+            .await
+    }
+}
+
+impl<'flox> Publish<'flox, NixAnalysis> {
+    /// copy the outputs and dependencies of the package to binary store
+    pub async fn upload_binary(self) -> PublishResult<Publish<'flox, NixAnalysis>> {
+        todo!()
+    }
+
+    #[allow(unused)] // until implemented
+    async fn get_binary_cache_metadata(
+        &self,
+        substituter: SubstituterUrl,
+    ) -> PublishResult<CacheMeta> {
+        todo!()
+    }
+
+    /// write snapshot to catalog and push to origin
+    pub async fn push_catalog(self) -> PublishResult<()> {
+        let url = self.publish_ref.clone_url();
+        let repo_dir = tempfile::tempdir_in(&self.flox.temp_dir)
+            .unwrap()
+            .into_path(); // todo catch error
+        let catalog = <Git as GitProvider>::clone(url, &repo_dir, false)
+            .await
+            .unwrap(); // todo: catch error
+
+        if catalog.list_branches().await.unwrap() // todo: catch error
+            .into_iter().any(|info| info.name == "catalog")
+        {
+            catalog.checkout("catalog", false).await.unwrap(); // todo: catch error
+        } else {
+            todo!();
+            // catalog.checkout("catalog", true).await.unwrap(); // todo: catch error
+
+            // catalog.set_upstream("origin", "catalog").await.unwrap();  // todo: implement
+            //                                                               todo: catch error
+        }
+        todo!()
+    }
+
+    /// read out the current publish state
+    pub fn analysis(&self) -> &Value {
+        self.analysis.deref()
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum PublishError {
+    #[error("Failed to load metadata for the package '{0}' in '{1}': {2}")]
+    DrvMetadata(AttrPath, PublishRef, NixCommandLineRunJsonError),
+
+    #[error("Failed to load metadata for flake '{0}': {1}")]
+    FlakeMetadata(PublishRef, NixCommandLineRunJsonError),
+}
+
+type PublishResult<T> = Result<T, PublishError>;
+
+/// Publishable FlakeRefs
+///
+/// `publish` modifies branches of the source repository.
+/// Thus we can only publish to repositories in (remote*) git repositories.
+/// This enum represents the subset of flakerefs we can use,
+/// so we can avoid parsing and converting flakerefs within publish.
+/// [GitRef<protocol::File>] should in most cases be resolved to a remote type.
+#[derive(PartialEq, Eq, Clone, Debug, Display)]
+pub enum PublishRef {
+    Ssh(GitRef<protocol::SSH>),
+    Https(GitRef<protocol::HTTPS>),
+    // File(GitRef<protocol::File>),
```

Comment:
It makes sense to leave it in for testing, otherwise testing publish requires an upstream repo, which is a _suboptimal_ test...

I'm trying to justify when you would want a local only publish, because they are error prone (immediately get invalid once the repo leaves your machine).

In the current publish we actually consider TWO flakerefs, `--build-url` which is the flake we're publishing and `--channel-url` which is the flake or repo we are publishing to.

The problem with this "publish anything anywhere" is that you can't track where the catalog entry for the package is and that catalogs can be made up from packages from different channels resulting in a separately developing package set.